### PR TITLE
:sparkles: NEW: DFBR for Registering Images

### PR DESCRIPTION
**Pull request by ruqayya**
This PR contains code for the DFBR method for registering pair of images.

### CODE REVIEW

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 1]
```diff
@@ -1112,6 +1151,18 @@ def estimate_bspline_transform(
     Returns:
         2D deformation transformation represented by a grid of control points.
 
+    Examples:
+        >>> from tiatoolbox.tools.registration.wsi_registration import (
+        >>>     estimate_bspline_transform, apply_bspline_transform
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:1156]:
```suggestion
        ...     estimate_bspline_transform, apply_bspline_transform
```

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 2]
```diff
@@ -1112,6 +1151,18 @@ def estimate_bspline_transform(
...
+        >>> )
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:1157]:
```suggestion
        ... )
```

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 3]
```diff
@@ -1112,6 +1151,18 @@ def estimate_bspline_transform(
...
+        >>> bspline_transform = estimate_bspline_transform(
+        >>>     fixed_gray_thumbnail, moving_gray_thumbnail, fixed_mask, moving_mask,
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:1159]:
```suggestion
        ...     fixed_gray_thumbnail, moving_gray_thumbnail, fixed_mask, moving_mask,
```

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 4]
```diff
@@ -1112,6 +1151,18 @@ def estimate_bspline_transform(
...
+        >>>     grid_space=50.0, sampling_percent=0.1,
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:1160]:
```suggestion
        ...     grid_space=50.0, sampling_percent=0.1,
```

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 5]
```diff
@@ -1112,6 +1151,18 @@ def estimate_bspline_transform(
...
+        >>> )
+        >>> bspline_registered_image = apply_bspline_transform(
+        >>>     fixed_thumbnail, moving_thumbnail, bspline_transform
+        >>> )
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:1164]:
```suggestion
        ... )
```

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 6]
```diff
@@ -1233,3 +1284,218 @@ def apply_bspline_transform(
...
+
+class AffineWSITransformer:
+    """Resampling regions from a whole slide image.
+
+    This class is used to resample tiles/patches from a whole slide image
+    using transformation.
+
+    Example:
+        >>> from tiatoolbox.tools.registration.wsi_registration import (
+        >>> AffineWSITransformer
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:1297]:
```suggestion
        ... AffineWSITransformer
```

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 7]
```diff
@@ -1233,3 +1284,218 @@ def apply_bspline_transform(
...
+        >>> )
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:1298]:
```suggestion
        ... )
```

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 8]
```diff
@@ -1233,3 +1284,218 @@ def apply_bspline_transform(
...
+        >>> tfm = AffineWSITransformer(wsi_reader, transform_level0)
+        >>> output = tfm.read_rect(location, size, resolution=resolution, units="level")
+
+    """
+
+    def __init__(self, wsi_reader: WSIReader, transform: np.ndarray) -> None:
+        """Initialize object.
+
+        Args:
+            wsi_reader (WSIReader):
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:1311]:
```suggestion
            img (WSIReader):
```

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 9]
```diff
@@ -1233,3 +1284,218 @@ def apply_bspline_transform(
...
+        >>> )
+        >>> from tiatoolbox.wsicore.wsireader import WSIReader
+        >>> wsi_reader = WSIReader.open(input_img=sample_ome_tiff)
+        >>> transform_level0 = np.eye(3)
+        >>> tfm = AffineWSITransformer(wsi_reader, transform_level0)
+        >>> output = tfm.read_rect(location, size, resolution=resolution, units="level")
+
+    """
+
+    def __init__(self, wsi_reader: WSIReader, transform: np.ndarray) -> None:
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:1307]:
```suggestion
    def __init__(self, img: WSIReader, transform: np.ndarray) -> None:
```

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 10]
```diff
@@ -1233,3 +1284,218 @@ def apply_bspline_transform(
...
+        """Initialize object.
+
+        Args:
+            wsi_reader (WSIReader):
+                An object with base WSIReader as base class.
+            transform (:class:`numpy.ndarray`):
+                A 3x3 transformation matrix.
+
+        """
+        self.wsi_reader = wsi_reader
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:1317]:
```suggestion
        self.wsi_reader = img
```

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 11]
```diff
@@ -1233,3 +1284,218 @@ def apply_bspline_transform(
...
+        """Compute patch size needed for transformation.
+
+        Args:
+            size (tuple(int)):
+                (width, height) tuple giving the desired output image size.
+            transform (:class:`numpy.ndarray`):
+                Transformation matrix of shape (3, 3).
+
+        Returns:
+            :tuple(int) - Maximum patch size needed for transformation.
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:1353]: Please can you check this? It doesn't seem right.

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 12]
```diff
@@ -1031,13 +1058,25 @@ def register(
...
+            moving_tissue_img, moving_tissue_mask = (
+                transform_tissue_img,
+                transform_tissue_mask,
+            )
+            before_dice = after_dice
+        else:
             block_transform = np.eye(3, 3)
 
+        # Fix translation offset
+        shift, _error, _diffphase = phase_cross_correlation(
```
Comment (shaneahmed) [inline tiatoolbox/tools/registration/wsi_registration.py:1073]:
```suggestion
        shift, _error, _diff_phase = phase_cross_correlation(
```

**Review by shaneahmed — CHANGES_REQUESTED** [review]
@ruqayya 

Please make changes as requested. Please also remove jupyter notebook and examples/README from this PR.

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 9 above)
Comment (John-P) [reply tiatoolbox/tools/registration/wsi_registration.py:1307]: I think 'reader' might be better. Calling it 'img' could be a little misleading.

**Review by John-P — COMMENTED** [review]
Can I just check why the notebook was deleted in this PR? Is this because it is now being done in #494?

File: tiatoolbox/tools/registration/wsi_registration.py [hunk 13]
```diff
@@ -1233,3 +1284,218 @@ def apply_bspline_transform(
...
+        ... )
+        >>> from tiatoolbox.wsicore.wsireader import WSIReader
+        >>> wsi_reader = WSIReader.open(input_img=sample_ome_tiff)
+        >>> transform_level0 = np.eye(3)
+        >>> tfm = AffineWSITransformer(wsi_reader, transform_level0)
+        >>> output = tfm.read_rect(location, size, resolution=resolution, units="level")
+
+    """
+
+    def __init__(self, wsi_reader: WSIReader, transform: np.ndarray) -> None:
```
Comment (John-P) [inline tiatoolbox/tools/registration/wsi_registration.py:1307]: Should this implement the WSIReader interface and be in wsireader? @shaneahmed

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 9 above)
Comment (shaneahmed) [reply tiatoolbox/tools/registration/wsi_registration.py:1307]: We are using img in model engines.

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 13 above)
Comment (shaneahmed) [reply tiatoolbox/tools/registration/wsi_registration.py:1307]: Let's keep it here for now. When you start working with Zarr we can move it and implement as WSIReader interface. What do you think?

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 9 above)
Comment (shaneahmed) [reply tiatoolbox/tools/registration/wsi_registration.py:1307]: @John-P  In the revised jupyter notebook you are using `reader`, so I am OK as far as we are consistent.

**Comment by shaneahmed** [comment]
> 

Yes to notebook is being separate from this PR to help David with review.

**Comment by shaneahmed** [comment]
> Can I just check why the notebook was deleted in this PR? Is this because it is now being done in #494?

Yes to notebook is being separate from this PR to help David with review.

**Comment by ruqayya** [comment]
> Can I just check why the notebook was deleted in this PR? Is this because it is now being done in #494?

Yes, Shan suggested it. The notebook is kept in #494 for review.

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 11 above)
Comment (shaneahmed) [reply tiatoolbox/tools/registration/wsi_registration.py]: Please use this format
https://github.com/TissueImageAnalytics/tiatoolbox/blob/9be4bf5aa03838e995e863c4425ed8b968c6c5f4/tiatoolbox/wsicore/wsireader.py#L473

Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py]: Updated.
https://github.com/TissueImageAnalytics/tiatoolbox/blob/faa20500c80e846f1ae5d8e62a8a36a8c832dbc6/tiatoolbox/tools/registration/wsi_registration.py#L1358-L1363

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 10 above)
Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py]: Updated.
https://github.com/TissueImageAnalytics/tiatoolbox/blob/faa20500c80e846f1ae5d8e62a8a36a8c832dbc6/tiatoolbox/tools/registration/wsi_registration.py#L1313-L1323

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 9 above)
Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py]: https://github.com/TissueImageAnalytics/tiatoolbox/blob/faa20500c80e846f1ae5d8e62a8a36a8c832dbc6/tiatoolbox/tools/registration/wsi_registration.py#L1313-L1323

File: tiatoolbox/tools/registration/wsi_registration.py (hunk 8 above)
Comment (ruqayya) [reply tiatoolbox/tools/registration/wsi_registration.py]: Renamed wsi_reader to reader.
https://github.com/TissueImageAnalytics/tiatoolbox/blob/faa20500c80e846f1ae5d8e62a8a36a8c832dbc6/tiatoolbox/tools/registration/wsi_registration.py#L1313-L1323

**Review by shaneahmed — APPROVED** [review]
Thanks @ruqayya

**Pull request merged**